httpx/metrics_test.go

```diff
@@ -166,88 +166,91 @@ func TestStressMetrics(t *testing.T) {
 	}
 
 	var (
-		httptimeout = 500 * time.Millisecond
 		testTimeout = 10 * time.Second
-		numSenders  = 4
+		httpTimeout = 50 * time.Millisecond
+
+		numSenders           = 4
+		numMessagesPerSender = 250
 
 		numExchangesSent int64 = 0
 		numTimeoutsSent  int64 = 0
-		ExSentMutex      sync.Mutex
-		TimeoutSentMutex sync.Mutex
+
+		timeoutCountCh = make(chan struct{})
+		messageCountCh = make(chan struct{})
+		errs           = make(chan error)
 	)
 
-	errs := make(chan error)
 	genctx, metricsctx, _ := session.NewContexts(&session.Config{
 		Debug:             false,
 		DebugHTTPMessages: false,
-		HTTPTimeout:       50 * time.Millisecond, // Short correlator time so we can actually get some timeouts.
+		HTTPTimeout:       httpTimeout, // Short correlator time so we can actually get some timeouts.
 	}, errs)
 
 	backend := newTestMetricsBackend()
 	metrics := &Metrics{
 		in:  make(chan interface{}),
 		met: backend,
 	}
+
 	go metrics.Init(metricsctx, nil)
 
 	for i := 0; i < numSenders; i++ {
+
 		go func() {
-			for {
-				select {
-				case <-genctx.StdContext.Done():
-					return
-				default:
-					connID := ksuid.New().String()
-					eID := ksuid.New().String()
-					timeout := rand.Intn(2)
-
-					if timeout == 0 {
-						metrics.In() <- &Request{
-							ConnectionID: connID,
-							ExchangeID:   eID,
-							Created:      refTime,
-							Method:       http.MethodGet,
-							URL:          newURL("/test"),
-						}
-						metrics.In() <- &Response{
-							ConnectionID: connID,
-							ExchangeID:   eID,
-							Created:      refTime.Add(10 * time.Millisecond),
-							StatusCode:   http.StatusOK,
-						}
-
-						ExSentMutex.Lock()
-						numExchangesSent++
-						ExSentMutex.Unlock()
-
-					} else {
-						metrics.In() <- &Request{
-							ConnectionID: connID,
-							ExchangeID:   eID,
-							Created:      refTime,
-							Method:       http.MethodGet,
-							URL:          newURL("/test"),
-						}
-
-						TimeoutSentMutex.Lock()
-						numTimeoutsSent++
-						TimeoutSentMutex.Unlock()
-					}
+			for msgCount := 0; msgCount < numMessagesPerSender; msgCount++ {
+				connID := ksuid.New().String()
```

Comment:
One thing we might want to consider doing is have this be configurable to simulate long-running connections as well.